### ISSUE TITLE
🐛 Bug Report: Techdocs Memory Leak

### ISSUE THREAD

**Issue by pcgqueiroz** [backstage/backstage#27347]
### 📜 Description

The `techdocs-backend` plugin is experiencing a memory leak when integrated with the `search-backend-module-techdocs`. In a production setup with a dedicated pod for `techdocs-backend`, the `/static/docs` endpoint is queried by the `search-backend-module-techdocs` every 10 minutes. With over 3000 entities to process, memory consumption increases progressively until the JavaScript heap is exhausted, resulting in a crash.

### 👍 Expected behavior

The `techdocs-backend` plugin should handle repeated `/static/docs` queries and large volumes of entities without excessive memory consumption or memory leaks, allowing stable operation in production.

### 👎 Actual Behavior with Screenshots

The memory usage of the `techdocs-backend` plugin gradually increases each time it processes requests from the `search-backend-module-techdocs`, eventually causing the JavaScript heap to overflow and crash.

![image](https://github.com/user-attachments/assets/24f66fda-2ca6-43b9-9966-353bedccfc39)

```log
<--- Last few GCs --->
[1:0x65c76b0] 20854057 ms: Scavenge 2021.6 (2074.2) -> 2019.2 (2076.2) MB, 11.9 / 0.0 ms  (average mu = 0.941, current mu = 0.519) allocation failure;
[1:0x65c76b0] 20854079 ms: Scavenge 2023.1 (2076.2) -> 2021.2 (2079.2) MB, 14.9 / 0.0 ms  (average mu = 0.941, current mu = 0.519) allocation failure;
[1:0x65c76b0] 20855608 ms: Scavenge 2026.6 (2079.2) -> 2024.2 (2097.2) MB, 1527.2 / 0.0 ms  (average mu = 0.941, current mu = 0.519) allocation failure;
<--- JS stacktrace --->
FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
 1: 0xb87bc0 node::Abort() [node]
 2: 0xa96834  [node]
 3: 0xd687f0 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [node]
 4: 0xd68b97 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [node]
 5: 0xf462a5  [node]
 6: 0xf5878d v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [node]
 7: 0xf32e8e v8::internal::HeapAllocator::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [node]
 8: 0xf34257 v8::internal::HeapAllocator::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [node]
 9: 0xf1542a v8::internal::Factory::NewFillerObject(int, v8::internal::AllocationAlignment, v8::internal::AllocationType, v8::internal::AllocationOrigin) [node]
10: 0x12da78f v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [node]
11: 0x170a079  [node]
```

### 👟 Reproduction steps

1. Deploy `techdocs-backend` in a microservices setup with a dedicated pod.
2. Configure it with a local builder/generator and AWS S3-type local storage.
3. Set up a regular polling interval (e.g., 10 minutes) for the `search-backend-module-techdocs` on `/static/docs`.
4. Observe memory usage as the `techdocs-backend` processes a large set of entities (over 3000, with some missing documentation in S3).

### 🖥️ Your Environment

**Backstage version:** 1.32.3
**Node.js version:** v18.18.0
**Setup:** Microservices with a dedicated pod for `techdocs-backend`
**TechDocs Configuration:** Local builder and generator, publishing files to a local AWS S3-type bucket

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

No, but I'm happy to collaborate on a PR with someone else

**Comment by awanlin** [comment]
Hi @pcgqueiroz, can you share your full Backstage details by running `yarn backstage-cli info` and sharing the results? I think that perhaps your are using the default [Lunr search engine](https://backstage.io/docs/features/search/search-engines#lunr) here which isn't ideal in production as it's entirely in-memory. We've had issues with it and the TechDocs Search Collator on the Demo site that we actually had to disable it, details on that are here: https://github.com/backstage/backstage/issues/23047.

We ship the [Postgres Search Engine](https://backstage.io/docs/features/search/search-engines#postgres) when creating a new Backstage instance as of a few releases ago to help avoid this as we also recommend using Postgres as your database for Production. I would highly suggest using this. 👍

**Comment by pcgqueiroz** [comment]
Hi @awanlin, I am already using the Postgres Search Engine `@backstage/plugin-search-backend-module-pg` in production.

Also, the search engine backend and their modules are in a different pod (which is not leaking memory), since my backstage in production runs in a micro-services architecture.

The pod that is leaking memory has only the `techdocs-backend` plugin and a dedicated `catalog-backend` plugin running.

Here is my configuration:

OS:   Linux 6.8.0-47-generic - linux/x64
node: v18.18.0
yarn: 4.5.0
cli:  0.28.1 (installed)
backstage:  1.32.3

Dependencies:
  @backstage/app-defaults                                            1.5.12
  @backstage/backend-app-api                                         0.7.9, 0.9.3, 1.0.1
  @backstage/backend-common                                          0.21.7, 0.23.3, 0.24.1, 0.25.0
  @backstage/backend-defaults                                        0.4.4, 0.5.2
  @backstage/backend-dev-utils                                       0.1.5
  @backstage/backend-openapi-utils                                   0.2.0
  @backstage/backend-plugin-api                                      0.6.21, 0.7.0, 0.8.1, 1.0.1
  @backstage/backend-tasks                                           0.5.27, 0.6.1
  @backstage/backend-test-utils                                      1.0.2
  @backstage/catalog-client                                          1.7.1
  @backstage/catalog-model                                           1.7.0
  @backstage/cli-common                                              0.1.14
  @backstage/cli-node                                                0.2.9
  @backstage/cli                                                     0.28.1
  @backstage/config-loader                                           1.9.1
  @backstage/config                                                  1.2.0
  @backstage/core-app-api                                            1.15.1
  @backstage/core-compat-api                                         0.3.1
  @backstage/core-components                                         0.14.10, 0.15.1
  @backstage/core-plugin-api                                         1.10.0
  @backstage/dev-utils                                               1.1.2
  @backstage/e2e-test-utils                                          0.1.1
  @backstage/errors                                                  1.2.4
  @backstage/eslint-plugin                                           0.1.10
  @backstage/frontend-app-api                                        0.10.0
  @backstage/frontend-defaults                                       0.1.1
  @backstage/frontend-plugin-api                                     0.9.0
  @backstage/frontend-test-utils                                     0.2.1
  @backstage/integration-aws-node                                    0.1.12
  @backstage/integration-react                                       1.2.0
  @backstage/integration                                             1.15.1
  @backstage/plugin-api-docs                                         0.11.11
  @backstage/plugin-app                                              0.1.1
  @backstage/plugin-auth-backend-module-atlassian-provider           0.3.1
  @backstage/plugin-auth-backend-module-auth0-provider               0.1.1
  @backstage/plugin-auth-backend-module-aws-alb-provider             0.2.1
  @backstage/plugin-auth-backend-module-azure-easyauth-provider      0.2.1
  @backstage/plugin-auth-backend-module-bitbucket-provider           0.2.1
  @backstage/plugin-auth-backend-module-bitbucket-server-provider    0.1.1
  @backstage/plugin-auth-backend-module-cloudflare-access-provider   0.3.1
  @backstage/plugin-auth-backend-module-gcp-iap-provider             0.3.1
  @backstage/plugin-auth-backend-module-github-provider              0.2.1
  @backstage/plugin-auth-backend-module-gitlab-provider              0.2.1
  @backstage/plugin-auth-backend-module-google-provider              0.2.1
  @backstage/plugin-auth-backend-module-guest-provider               0.2.1
  @backstage/plugin-auth-backend-module-microsoft-provider           0.2.1
  @backstage/plugin-auth-backend-module-oauth2-provider              0.3.1
  @backstage/plugin-auth-backend-module-oauth2-proxy-provider        0.2.1
  @backstage/plugin-auth-backend-module-oidc-provider                0.3.1
  @backstage/plugin-auth-backend-module-okta-provider                0.1.1
  @backstage/plugin-auth-backend-module-onelogin-provider            0.2.1
  @backstage/plugin-auth-backend                                     0.23.1
  @backstage/plugin-auth-node                                        0.4.17, 0.5.3
  @backstage/plugin-auth-react                                       0.1.7
  @backstage/plugin-bitbucket-cloud-common                           0.2.24
  @backstage/plugin-catalog-backend-module-github                    0.7.6
  @backstage/plugin-catalog-backend-module-ldap                      0.9.1
  @backstage/plugin-catalog-backend-module-logs                      0.1.3
  @backstage/plugin-catalog-backend-module-openapi                   0.2.3
  @backstage/plugin-catalog-backend-module-scaffolder-entity-model   0.2.1
  @backstage/plugin-catalog-backend-module-unprocessed               0.5.1
  @backstage/plugin-catalog-backend                                  1.27.1
  @backstage/plugin-catalog-common                                   1.1.0
  @backstage/plugin-catalog-graph                                    0.4.11
  @backstage/plugin-catalog-import                                   0.12.5
  @backstage/plugin-catalog-node                                     1.13.1
  @backstage/plugin-catalog-react                                    1.14.0
  @backstage/plugin-catalog-unprocessed-entities-common              0.0.4
  @backstage/plugin-catalog-unprocessed-entities                     0.2.9
  @backstage/plugin-catalog                                          1.24.0
  @backstage/plugin-devtools-backend                                 0.4.1
  @backstage/plugin-devtools-common                                  0.1.12
  @backstage/plugin-devtools                                         0.1.19
  @backstage/plugin-events-node                                      0.3.10, 0.4.3
  @backstage/plugin-home-react                                       0.1.18
  @backstage/plugin-home                                             0.8.0
  @backstage/plugin-org                                              0.6.31
  @backstage/plugin-permission-backend                               0.5.50
  @backstage/plugin-permission-common                                0.7.14, 0.8.1
  @backstage/plugin-permission-node                                  0.7.32, 0.8.4
  @backstage/plugin-permission-react                                 0.4.27
  @backstage/plugin-scaffolder-backend-module-azure                  0.2.1
  @backstage/plugin-scaffolder-backend-module-bitbucket-cloud        0.2.1
  @backstage/plugin-scaffolder-backend-module-bitbucket-server       0.2.1
  @backstage/plugin-scaffolder-backend-module-bitbucket              0.3.1
  @backstage/plugin-scaffolder-backend-module-confluence-to-markdown 0.3.1
  @backstage/plugin-scaffolder-backend-module-gerrit                 0.2.1
  @backstage/plugin-scaffolder-backend-module-gitea                  0.2.1
  @backstage/plugin-scaffolder-backend-module-github                 0.5.1
  @backstage/plugin-scaffolder-backend-module-gitlab                 0.6.0
  @backstage/plugin-scaffolder-backend                               1.26.2
  @backstage/plugin-scaffolder-common                                1.5.6
  @backstage/plugin-scaffolder-node                                  0.4.11, 0.5.0
  @backstage/plugin-scaffolder-react                                 1.13.2
  @backstage/plugin-scaffolder                                       1.26.2
  @backstage/plugin-search-backend-module-catalog                    0.2.4
  @backstage/plugin-search-backend-module-explore                    0.2.4
  @backstage/plugin-search-backend-module-pg                         0.5.37
  @backstage/plugin-search-backend-module-techdocs                   0.3.1
  @backstage/plugin-search-backend-node                              1.3.4
  @backstage/plugin-search-backend                                   1.6.1
  @backstage/plugin-search-common                                    1.2.14
  @backstage/plugin-search-react                                     1.8.1
  @backstage/plugin-search                                           1.4.18
  @backstage/plugin-signals-node                                     0.1.13
  @backstage/plugin-signals-react                                    0.0.6
  @backstage/plugin-techdocs-backend                                 1.11.1
  @backstage/plugin-techdocs-common                                  0.1.0
  @backstage/plugin-techdocs-module-addons-contrib                   1.1.16
  @backstage/plugin-techdocs-node                                    1.12.12
  @backstage/plugin-techdocs-react                                   1.2.9
  @backstage/plugin-techdocs                                         1.11.0
  @backstage/plugin-user-settings-backend                            0.2.26
  @backstage/plugin-user-settings-common                             0.0.1
  @backstage/plugin-user-settings                                    0.8.14
  @backstage/release-manifests                                       0.0.11
  @backstage/test-utils                                              1.7.0
  @backstage/theme                                                   0.5.7, 0.6.0
  @backstage/types                                                   1.1.1
  @backstage/version-bridge                                          1.0.10

**Comment by iamEAP** [comment]
Hey there!  I happened to be investigating something similar and happened upon this thread.

@pcgqueiroz, I believe the problem you're experiencing may be specific to the AWS publisher implementation, related to [this line in the TechDocs awsS3 publisher](https://github.com/backstage/backstage/blob/d995579/plugins/techdocs-node/src/stages/publish/awsS3.ts#L480).

Other implementations of the docsRouter will pipe the stream returned from the object store service (e.g. [here in the GCS implementation](https://github.com/backstage/backstage/blob/fbdc631/plugins/techdocs-node/src/stages/publish/googleStorage.ts#L342)), but the AWS implementation seems to be loading the entire file contents into memory before sending it on to clients.  

I unfortunately don't have the bandwidth (nor a nice/easy test environment) to fix this, but my assumption is that removing the `streamToBuffer()` call and instead piping the stream into the response will reduce memory usage substantially.

**Comment by squid-ney** [comment]
Hi, do you have TechDocs Cache set-up? I'm trying to replicate this issue.

**Comment by pcgqueiroz** [comment]
Hey Sydney,

I don´t have techdocs cache set-up. Please tell me if you have any other
questions so you can replicate the issue.

**Comment by pcgqueiroz** [comment]
Hey @iamEAP,

I have tried to make the changes you have suggested

```diff
- res.send(await streamToBuffer(resp.Body as Readable));
+ const fileStream = resp.Body as Readable;
+ fileStream.pipe(res);
```

but unfortunately it did not work.

Any other ideas that I could test? Thanks

**Comment by Ferin79** [comment]
### Memory Profiling Snapshot (1 Week)

The following memory profiling snapshot from DataDog indicates a potential memory leak:

![Memory Profiling Screenshot](https://github.com/user-attachments/assets/dd344d95-a44b-4d65-90e6-a331d900b569)

It appears that `keyv` is leaking memory. This issue was addressed in version `4.3.0` as noted in [this GitHub issue](https://github.com/jaredwray/keyv/issues/101).

Currently, Backstage is using an older version (`4.0.0`) of `keyv`, hoisted from the following dependency chain:  
`_project_#backend#@backstage#plugin-scaffolder-backend#@backstage#plugin-scaffolder-backend-module-gitlab#@gitbeaker#node#got#cacheable-request#keyv`.

**Comment by squid-ney** [comment]
Hi @pcgqueiroz and @Ferin79!
I'm digging into this issue more since the suggested change to pipe the stream did not work. @pcgqueiroz could you share how you confirmed the piping did not work?  

@Ferin79 Thanks for pointing out the need to upgrade `keyv`! I'm cautious to say that the `keyv` library is causing this specific issue though. It looks like `keyv` is only used in TechDocs for the TechDocs cache and @pcgqueiroz's TechDocs set-up is not using cache. We should still update this dependency though.

**Comment by squid-ney** [comment]
I think this could be related to the `S3Client` configuration. I was reading through this issue, https://github.com/aws/aws-sdk-js-v3/issues/3560, and noticed some discussion on the socket settings for v3. Maybe playing around with these socket settings could help the memory usage by setting a max for socket growth and making sure there is a timeout so they can be cleaned up.

**Comment by pcgqueiroz** [comment]
Hi @squid-ney, thanks for looking at this issue. Answering your question, when I replaced the buffer by the pipe, the code did work but it didn't solve the memory leak problem. Sorry that I was not enough clear in my comment above.

In my tests, I have the same feeling that the issue is related to the `S3Client`. Let me know if I can help you in anyway.

**Comment by iamEAP** [comment]
Do you happen to have the permission framework enabled?

There's one more code path in TechDocs that will cause `keyv` to be used in that case (in order to cache permission responses for a given user and entity).

**Comment by pcgqueiroz** [comment]
Hey @iamEAP , I do not know exactly what you mean by having the permission framework enabled, but I do have the `@backstage/plugin-permission-backend` enabled.

The `techdocs.cache` option is not enabled/configured.

**Comment by iamEAP** [comment]
Yes, this wouldn't respect `techdocs.cache`.  Basically, when `permissions.enabled` is set to `true` in `app-config.yaml`, [this middleware](https://github.com/backstage/backstage/blob/5edd344/plugins/techdocs-backend/src/service/router.ts#L300-L327) is applied to ensure that anyone loading docs content for a given entity has permission to _view_ that entity.  It does that check by attempting to load the entity before serving any content associated with it.

This check could happen _frequently_ (because it would be called not just for the HTML content of a page, but also all images/assets loaded on that page), so the result [gets cached](https://github.com/backstage/backstage/blob/5edd344/plugins/techdocs-backend/src/service/CachedEntityLoader.ts#L54).

All of that to say: don't rule the `keyv` issue out entirely.  Hopefully it's straightforward to bump that dep in your `yarn.lock`.  Something quick and easy to rule out.

**Comment by pcgqueiroz** [comment]
As far as I could trace in my `yarn.lock` file, the only `keyv` version with memory leak is `4.0.0`, and it is only related to the package `@backstage/plugin-scaffolder-backend-module-gitlab` which I am not using.

Unfortunately I do not thing this is related :-(

**Comment by Ferin79** [comment]
@pcgqueiroz  The package `@backstage/plugin-scaffolder-backend-module-gitlab` is being installed by `@backstage/plugin-scaffolder-backend`. So even if you are not using it, it's being installed by scaffolder backend.

Also the `got` library has leak when using `cache`. https://github.com/sindresorhus/got/issues/1128

**Comment by squid-ney** [comment]
Hey @pcgqueiroz a few questions about your TechDocs set-up.
- In the issue description, you mention using an "AWS S3-type local storage", could you elaborate on this more? 
- Are you setting a custom `parallelismLimit`  for the techdocs collator in `search-backend-module-techdocs`?

**Comment by pcgqueiroz** [comment]
Hi @squid-ney, answering your first question, using an "aws s3-type local storage" means that I am using `awsS3` as the `techdocs.publisher.type` configuration (and I have an on-prem setup for bucket storage).

I am not setting a custom `parallelismLimit` for the techdocs collator. Should I?

**Comment by drodil** [comment]
We are also experiencing OOM issues and apparently it's the TechDocs causing this. We took heap dumps out of the system when it was stable and also after it started munching the memory. And it seems the root cause is in the `DocsSynchronizer`. Here's a screenshot of the comparison:

![Image](https://github.com/user-attachments/assets/eaa46e95-0454-4a12-adb5-ea12d4871b1d)

There are over 200k same size `JSArrayBufferData` objects in heap.

**Comment by drodil** [comment]
Just a hunch that the response socket during TechDocs sync does not get closed for some reason (https://github.com/backstage/backstage/blob/master/plugins/techdocs-backend/src/service/router.ts#L347). Tried to check where this could happen but couldn't find anything suspicious. For context, we are using the local builder and generator and publishing to S3.

**Comment by drodil** [comment]
I was wondering if it would make sense to have some sort of timeout in the `EventStream` such as if there's no action in, let's say 30 seconds, it will forcibly close the socket:

```diff
diff --git a/plugins/techdocs-backend/src/service/router.ts b/plugins/techdocs-backend/src/service/router.ts
index b99579e0eb4..62def812b92 100644
--- a/plugins/techdocs-backend/src/service/router.ts
+++ b/plugins/techdocs-backend/src/service/router.ts
@@ -354,8 +354,26 @@ export function createEventStream(
     'Content-Type': 'text/event-stream',
   });
 
+  let timeout;
+
+  const stopTimeout = () => {
+    if (timeout) {
+      clearTimeout(timeout);
+    }
+  };
+
+  const startTimeout = () => {
+    stopTimeout();
+    timeout = setTimeout(() => {
+      res.end();
+    }, 30000);
+  };
+
+  startTimeout();
+
   // client closes connection
   res.socket?.on('close', () => {
+    stopTimeout();
     res.end();
   });
 
@@ -372,15 +390,18 @@ export function createEventStream(
   return {
     log: data => {
       send('log', data);
+      startTimeout();
     },
 
     error: e => {
       send('error', e.message);
+      stopTimeout();
       res.end();
     },
 
     finish: result => {
       send('finish', result);
+      stopTimeout();
       res.end();
     },
   };

```
Also, it probably should be beneficial to have some maximum time to wait for the docs to build in `DocsSynchronizer` (where it's logging `The docs building process is taking a little bit longer to process this entity. Please bear with us`). In our case, the container memory goes up from ~2GB to 10GB in approximately 1 hour, 40 minutes so the maximum time could be even quite high. @freben WDYT?

**Comment by freben** [comment]
It'd be great if you had the chance to yarn-patch in a minimal change and just see the problem not reoccur - if you can

**Comment by squid-ney** [comment]
This week some @backstage/techdocs-maintainers team members were investigating a different Techdocs indexing issue internally and we now see the same errors `(node:3233) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 close listeners added to [Socket] Use emitter.setMaxListeners() to increase limit` in our techdocs-backend logs. We are using the google storage configuration. We think numerous sockets  are being opened and never closed causing odd behavior in our system.

At the moment, we are exploring if this issue could be related to the configuration of the nginx server that is running our techdocs-backend. In past attempts to replicate this issue, I set up a test AWS configuration on my local and ran some intense load tests against the Techdocs backend, but was never able to replicate the memory spikes. Think I'll try again to replicate by running the backend in docker. 

We noticed our nginx configuration has a pool of 64 keepalive connections, anything similar in your set-up?

**Comment by drodil** [comment]
@squid-ney I think this comment might be related: https://github.com/aws/aws-sdk-js-v3/issues/3560#issuecomment-1474275964

**Comment by drodil** [comment]
We enabled `--trace-warnings` in our instance and there could be some pointers to the `MaxListnersExceededWarning`s:

![Image](https://github.com/user-attachments/assets/4bdb9b4f-c3e5-47f4-b878-6dd57226a9ee)

This probably should be investigated a bit more.

**Comment by drodil** [comment]
It seems to point to this: https://github.com/backstage/backstage/blob/961861292ee6f93c4948c6b5a16458fa8fa7b671/plugins/techdocs-backend/src/cache/cacheMiddleware.ts

I wonder if the socket gets closed properly in case the underlying storage returns an error here. We will try now disabling the TechDocs cache and see if it has any effect.

**Comment by drodil** [comment]
Small update: we disabled the TechDocs cache and for now it seems that it is not running out of memory anymore. Also, the `MaxListenersExceededWarning`s are gone. We will continue monitoring the situation over the weekend but it looks promising. Previously we were getting a lot of errors from the Techdocs collator: `TechDocs S3 router failed to serve static files from bucket...` which might be the root cause. So my guess of the root cause is:

1. TechDocs search collator tries to find `/static/docs/entity/search_index.json`
2. Also the `DocsSynchronizer` uses the `/static/docs/entity/techdocs_metadata.json` to check for the last build timestamp. That's why the heap snapshots showed so many sockets opened in this file.
3. If the S3 does not have the static files available, it will return a 404 error (the error above)
4. `cacheMiddleware` does not close the socket properly in the 404 (or any other error) case

WDYT @freben, does this make any sense?

**Comment by freben** [comment]
Pointing @squid-ney I think since I'm a bit out of my depth!

**Comment by alexlorenzi** [comment]
Hey @drodil  - So, if removing the cache resolves the `MaxListener...` issues then it's likely the cause originated [here](https://github.com/backstage/backstage/blob/4de4602e941397dfd6129219dec69d89e80ff17c/plugins/techdocs-backend/src/cache/cacheMiddleware.ts#L70-L81), which is similar to an issue that we faced internally (that @squid-ney mentioned). As you pointed out, basically the sockets aren't getting closed straight away (due to some `keep-alive` configuration we have in our case) so if there's a lot of load (in times of indexing for example) lots of listeners will likely be opened up on the same socket.

I've been working on a fix that uses a map to handle the cache write so that only a single listener is opened per socket, and it seems to help. I'll open a PR soon, and I'll link it here so you can see it too.

We also had some temporary relief by increasing the number of pods for our Techdoc Backend service which helped spread the load

**Comment by drodil** [comment]
Hi @alexlorenzi 

That indeed might be the source but I will get back to report on Monday how our instance has been running over the weekend.

I think it would be beneficial to add unit test here: https://github.com/backstage/backstage/blob/master/plugins/techdocs-backend/src/cache/cacheMiddleware.test.ts#L68 that sends the 404 response like for example S3 storage does and see what happens. We should get some reproducible way of proving this before a fix IMO.

**Comment by drodil** [comment]
Update after the weekend; it indeed stopped munching the memory after removing the cache middleware and now it does not kill containers because of OOM. The only issue now is that it seems that the TechDocs indexer task gets a timeout by the Janitor every 30 minutes even if the config has a timeout of 2 hours. Have to check what is going on in the database for this.

**Comment by drodil** [comment]
@jonesbusy do you also have the techdocs cache enabled? The root cause is there but have not had the time to debug or reproduce this further.

**Comment by alexlorenzi** [comment]
@drodil 
I created this draft PR that aims to address the issue of continually adding listeners to sockets that are kept open, it seems to help though I'm still testing. 

I don't know if you're able to patch it in, but I'd be interested to hear if it helps with your issue too.

https://github.com/backstage/backstage/pull/28833/files

**Comment by drodil** [comment]
@alexlorenzi cool, have to read that PR through. On a side note, there might be some ideas in the battle-tested `http-cache-middleware`: https://github.com/BackendStack21/http-cache-middleware/blob/master/index.js and https://github.com/BackendStack21/on-http-end/blob/master/index.js

**Comment by zcmander** [comment]
^ I was looking at the `http-cache-middleware` express plugin today morning as well and left to wonder why do we even need to have a custom middleware for caching. This sounds something that could work with off-the-shelf plugin such as `http-cache-middleware`. And cherry on the top: it support redis quite nicely (well, not with Backstage's cache service obviously but it shouldn't be too hard to add).

As we've seen: this is quite error-prone and may lead to hard-to-diagnose errors, so I would vote for something more robust and battle-tested solution. @alexlorenzi thank you for the effort for fixing this. if nothing else, this can be used to fix the issue in short term.

**Comment by drodil** [comment]
@zcmander just my thoughts. It could be made to `httpRouterServiceFactory` as another middleware by utilizing directly `http-cache-middleware` or by "forking" it and modifying it to utilize the `CacheService`. An additional win with this approach would be that the cache could be used in other plugins as well just by adding the necessary headers to the response; if needed. @freben WDYT?

**Comment by drodil** [comment]
Coming back to this, our instance is again munching the memory but not to OOM. The cache is disabled but the memory consumption rocketed over 3GB after the TechDocs search collator started. In the logs there are a lot of `TechDocs S3 router failed to serve static files from bucket` error messages. Something is again pointing out that it might not be the TechDocs cache but instead related to S3 or the search collator. Any additional ideas are more than welcome!

**Comment by squid-ney** [comment]
Looks like there are two separate issues here:
1. The TechDocs cache does not close sockets properly. Possible solution: 
- Look into replacing the custom cache middleware with `http-cache-middleware`.
- Temporarily increase the number of pods for techdocs-backend.
2. S3Client specific issue during TechDocs collation. Possible solutions: 
- Destroy and recreate the S3Client each time.
- Configure the S3 `maxSockets` and `socketTimeout` parameters.

**Comment by squid-ney** [comment]
@drodil Also, would be good to test this replace S3 streamToBuffer with pipe change: #28993 also to see if it improves issues during collation.

**Comment by Speissi** [comment]
Hi! Any updates to this issue? We are seeing similar behaviour even though we are using GCS instead of S3. Looking back at the our memory usage metrics and correlating the start of memory leak with commits and deployments, it seems like this started when we first enabled `permission`.

Based on that I did some quick tests and if we deploy Backstage with permission disabled sure enough the memory leak seems to stop.

Not sure if this is useful or even if the same issue, but please let me know if I can help.

**Comment by melloc01** [comment]
> We are seeing similar behaviour even though we are using GCS instead of S3.

Same here. We don't know when it started but we don't use S3 as well.

**Comment by vinzscam** [comment]
cc @backstage/techdocs-maintainers

**Comment by Vity01** [comment]
unstale

**Comment by Speissi** [comment]
Hello all, I think after some digging I've found the reason why TechDocs leaks memory in my case. If there's any mistakes in my understanding please let me know.

TL;DR: The cache key includes a fresh token per request, so every key is unique. If you are using the in-memory caching then expiration cleanup only occurs on access, since keys are never reused, expired entries pile up, causing memory bloat/OOM.

If the permission is enabled in app-config.yml then [CachedEntityLoader](https://github.com/backstage/backstage/blob/v1.40.2/plugins/techdocs-backend/src/service/router.ts#L308) will be used to load the entity. Cached entity loader then creates the cache key from entity name and the token provided. If the matching cache key is not in the cache then it will be cached with a TTL of 5 seconds.

For caching keyv is used and in app-config.yml you can configure different backends like Redis, Memcache and other for it. However if nothing is defined then [an in-memory map will be used](https://github.com/backstage/backstage/blob/v1.40.2/packages/backend-defaults/src/entrypoints/cache/CacheManager.ts#L319-L329) to store the cached entities. A quirk with the map is that expired entries are [not removed until they are accessed after the TTL has expired](https://github.com/jaredwray/keyv/blob/main/packages/keyv/src/index.ts#L536-L540).

This causes a problem with this particular use case since the cache key has the token in it and the [token is generated on every request](https://github.com/backstage/backstage/blob/v1.40.2/plugins/techdocs-backend/src/service/router.ts#L303-L306) meaning that the cache key will always be unique and we never trigger the TTL expiration cleanup in the map. Causing the memory to grow until the heap runs out of memory.

So I guess this is more of a "feature" and not a bug? The fix would be stop using map as cache backend and start using Redis or one of the other options, which solves this problem by having proper TTL handling and removing the need of an in-memory map of the cache.

Not sure if it would be helpful to include this as a part of documentation and included as part of rationale why Redis or some other proper caching backend should be used for anything other than experimentation.

PS: If I have indeed understood the cache key generation correctly then it seem there is really no benefit of caching anything here since the token thus the key is different every time? Looking at the implementation I can see that at [some point](https://github.com/backstage/backstage/commit/f1918143e51b3d261408794362f72df5714c188e) of time it was implemented so that the user identity is checked and  the user entity ref was used as cache key instead of the token. I wonder why this was changed?

PPS: If above is indeed correct then I'm more than happy to contribute to improving the logic. Just let me know.
PPPS: Opened a PR for the cache key https://github.com/backstage/backstage/pull/30652

**Comment by vinzscam** [comment]
ok, it seems like https://github.com/backstage/backstage/pull/30652 is out. Let us know if the issue is fixed

**Comment by jboeijenga** [comment]
For us the issue is indeed fixed after upgrading to 1.42.0. Thanks a lot!

<img width="852" height="254" alt="Image" src="https://github.com/user-attachments/assets/2262782d-a906-4d1e-8e58-733851dc85a2" />

**Comment by jonesbusy** [comment]
I will wait few days more, but also looking at my graph (nodejs old space) the memory usage is very different since the 1.42 deployment. So I'm pretty sure the issue is indeed fixed. Thanks!!

<img width="1153" height="583" alt="Image" src="https://github.com/user-attachments/assets/1e3ed510-e64b-4e8b-b3f4-d25cc80761c9" />

**Comment by pcgqueiroz** [comment]
Indeed, it seems the new version 1.42 fixed the problem! Thanks to everyone who followed this issue and to @Speissi who fixed it :-)

<img width="1108" height="214" alt="Image" src="https://github.com/user-attachments/assets/f5bc6b92-9393-4761-836b-41c5ea11949a" />
